### ISSUE TITLE
Object Comparisons in Python 3?

### ISSUE THREAD

**Issue by cathalgarvey** [dgrunwald/rust-cpython#90]
So, I just found an obscure doc-bug: It seems that the `ObjectProtocol.compare` method is [only defined for Python 2.7](https://github.com/dgrunwald/rust-cpython/blob/1129131501ccc8b64f78102a5d098dab8cacef5d/src/objectprotocol.rs#L75)?

I have two objects I'd like to compare; in Python-land they are comparable for equality. How do I do this in a version-agnostic way? I tried comparing the Rust objects, equivalent to `PyObject1 == PyObject2`, and this didn't work out; it compiled, but the results were not what I'd hoped. :)

**Comment by cathalgarvey** [comment]
Update: I've fallen back on comparing `.hash` return values, but that's probably not ideal in the general case where `__cmp__` would have applied.

**Comment by dgrunwald** [comment]
The `cmp` builtin and the `__cmp__` special method does not exist in Python 3.

We'll need to expose `PyObject_RichCompare` somewhere in the API.

I guess the API could look like this:
```
    fn eq<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
    fn ne<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
    fn lt<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
    fn gt<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
    fn le<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
    fn ge<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
```

The return type would be `PyObject`, because `__eq__` is allowed to return types other than `bool`.
So you'd end up with `obj1.eq(py, obj2)?.is_true()?`.

Though sometimes an `Ordering` value would be nice.. so I guess we could emulate the Python 2 `cmp()` logic on Python 3 by trying `eq`/`lt`/`gt`.

**Comment by cathalgarvey** [comment]
I think, Python-being-Python, that `obj1.eq(py, obj2)?.is_true()?` is reasonable. Also, that interface can be easily backported to the Py2 `cmp` approach, and a label attached suggesing it for version-agnosticism.

On 12 March 2017 12:09:38 GMT+00:00, Daniel Grunwald <notifications@github.com> wrote:
>The `cmp` builtin and the `__cmp__` special method does not exist in
>Python 3.
>
>We'll need to expose `PyObject_RichCompare` somewhere in the API.
>
>I guess the API could look like this:
>```
>    fn eq<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
>    fn ne<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
>    fn lt<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
>    fn gt<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
>    fn le<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
>    fn ge<O>(&self, py: Python, other: O) -> PyResult<PyObject>;
>```
>
>The return type would be `PyObject`, because `__eq__` is allowed to
>return types other than `bool`.
>So you'd end up with `obj1.eq(py, obj2)?.is_true()?`.
>
>Though sometimes an `Ordering` value would be nice.. so I guess we
>could emulate the Python 2 `cmp()` logic on Python 3 by trying
>`eq`/`lt`/`gt`.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/dgrunwald/rust-cpython/issues/90#issuecomment-285940404

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by dgrunwald** [comment]
Fixed in 51f7aba1a43eeb202202a9efb7bbebb3704cd2bc.

Instead of 6 methods, I added a single method `rich_compare` taking a `CompareOp` enum.
Both `compare` and `rich_compare` are now available independently of the Python version.

**Comment by cathalgarvey** [comment]
Wow, fast! Thank you. :)

On 12 March 2017 13:15:59 GMT+00:00, Daniel Grunwald <notifications@github.com> wrote:
>Fixed in 51f7aba1a43eeb202202a9efb7bbebb3704cd2bc.
>
>Instead of 6 methods, I added a single method `rich_compare` taking a
>`CompareOp` enum.
>Both `compare` and `rich_compare` are now available independently of
>the Python version.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/dgrunwald/rust-cpython/issues/90#issuecomment-285943895

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
